File: scavenger-agent-python/scavenger/internal/scheduler.py
import logging
from datetime import datetime, timedelta
from typing import Optional

from apscheduler.schedulers.background import BackgroundScheduler
from grpc import StatusCode
from grpc._channel import _InactiveRpcError

from scavenger.config import Config
from scavenger.internal.client import Client
from scavenger.internal.invocation_registry import InvocationRegistry
from scavenger.internal.model import SchedulerState
from scavenger.internal.scan import CodeBaseScanner
from scavenger.model.CodeBasePublication_pb2 import CodeBasePublication
from scavenger.model.InvocationDataPublication_pb2 import InvocationDataPublication

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.DEBUG)


class Scheduler:
    def __init__(self, config: Config, client: Client, codebase_scanner: CodeBaseScanner, invocation_registry: InvocationRegistry):
        self.config = config
        self.client = client
        self.codebase_scanner = codebase_scanner
        self.invocation_registry = invocation_registry

        self.scheduler = BackgroundScheduler(daemon=False)
        self.scheduler.add_job(self.run,
                               "interval",
                               seconds=self.config.scheduler_interval_seconds,
                               next_run_time=datetime.now() + timedelta(seconds=self.config.scheduler_initial_delay_seconds))
        self.is_code_base_published = False
        self.code_base_publication: Optional[CodeBasePublication] = None
        self.invocation_data_publication: Optional[InvocationDataPublication] = None
        self.code_base_fingerprint = None
        self.dynamic_config = None
        interval_seconds = 600
        retry_interval_seconds = 600
        self.force_interval_seconds = self.config.force_interval_seconds
        self.poll_state = SchedulerState("config_pool").initialize(interval_seconds, retry_interval_seconds)
        self.code_base_publisher_state = SchedulerState("codebase").initialize(interval_seconds, retry_interval_seconds)
        self.invocation_data_publisher_state = SchedulerState("invocation_data").initialize(interval_seconds, retry_interval_seconds)

        if self.force_interval_seconds != 0:
            self.poll_state.update_intervals(self.force_interval_seconds, self.force_interval_seconds)
            self.code_base_publisher_state.update_intervals(self.force_interval_seconds, self.force_interval_seconds)
            self.invocation_data_publisher_state.update_intervals(self.force_interval_seconds, self.force_interval_seconds)

    def start(self):
        logging.info("Scheduler is starting")
        self.scheduler.start()

    def run(self):
        logging.debug("Task is starting")
        self.poll_dynamic_config_if_needed()
        self.publish_code_base_if_needed()
        self.publish_invocation_data_if_needed()
        logging.debug("Task is done")

    def shutdown(self):
        self.scheduler.shutdown()
        if self.dynamic_config is not None:
            self.code_base_publisher_state.schedule_now()
            self.publish_code_base_if_needed()

            self.invocation_data_publisher_state.schedule_now()
            self.publish_invocation_data_if_needed()

    def poll_dynamic_config_if_needed(self):
        if self.poll_state.is_due_time():
            logging.debug("Poll dynamic config is due time.")
            # noinspection PyBroadException
            try:
                self.dynamic_config = self.client.poll_config()
                if self.force_interval_seconds == 0:
                    self.poll_state.update_intervals(
                        self.dynamic_config.config_poll_interval_seconds,
                        self.dynamic_config.config_poll_retry_interval_seconds
                    )
                    self.code_base_publisher_state.update_intervals(
                        self.dynamic_config.code_base_publisher_check_interval_seconds,
                        self.dynamic_config.code_base_publisher_retry_interval_seconds
                    )
                    self.invocation_data_publisher_state.update_intervals(
                        self.dynamic_config.invocation_data_publisher_interval_seconds,
                        self.dynamic_config.invocation_data_publisher_retry_interval_seconds
                    )
                self.poll_state.schedule_next()
            except Exception as e:
                if isinstance(e, _InactiveRpcError):
                    if e.code() == StatusCode.UNAUTHENTICATED:
                        logging.warning("Authentication is Failed.")
                        self.shutdown()
                logging.warning(e)
                self.poll_state.schedule_retry()

    def publish_code_base_if_needed(self):
        if not self.is_code_base_published and self.code_base_publisher_state.is_due_time() and self.dynamic_config is not None:
            logging.debug("Publish codebase is due time.")
            try:
                codebase = self.codebase_scanner.scan()

                if len(codebase.functions) > 100_000:
                    logging.warning("maximum methods count(100000) exceed", len(codebase.functions))
                    return

                if len(codebase.functions) == 0:
                    logging.warning("no methods are found")
                    return

                self.code_base_publication = CodeBasePublication(
                    common_data=self.config.build_common_publication_data(codebase.get_fingerprint(self.config)),
                    entry=[function.to_codebase_entry() for function in codebase.functions]
                )

                self.client.send_codebase_publication(self.code_base_publication)
                self.is_code_base_published = True
                self.code_base_fingerprint = self.code_base_publication.common_data.code_base_fingerprint
                self.code_base_publication = None
            except Exception as e:
                logging.warning(e)
                self.code_base_publisher_state.schedule_retry()

    def publish_invocation_data_if_needed(self):
        if self.invocation_data_publisher_state.is_due_time() and self.dynamic_config is not None and self.is_code_base_published:
            logging.debug("Publish invocation data is due time.")
            try:
                if self.invocation_data_publication is None:
                    invocations, recording_interval_started_at_millis = self.invocation_registry.get_invocations()
                    self.invocation_data_publication = InvocationDataPublication(
                        common_data=self.config.build_common_publication_data(self.code_base_fingerprint),
                        entry=[InvocationDataPublication.InvocationDataEntry(hash=hash_) for hash_ in invocations],
                        recording_interval_started_at_millis=recording_interval_started_at_millis
                    )
                self.client.send_invocation_data_publication(
                    self.invocation_data_publication
                )
                self.invocation_data_publication = None
                self.invocation_data_publisher_state.schedule_next()
            except Exception as e:
                logging.warning(e)
                self.invocation_data_publisher_state.schedule_retry()

File: scavenger-agent-java/src/main/java/com/navercorp/scavenger/javaagent/scheduling/Scheduler.java
package com.navercorp.scavenger.javaagent.scheduling;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.logging.Level;

import com.navercorp.scavenger.javaagent.collecting.CallStackRegistry;
import com.navercorp.scavenger.javaagent.collecting.InvocationRegistry;
import com.navercorp.scavenger.model.CallStackDataPublication;

import io.grpc.Status;
import io.grpc.StatusRuntimeException;
import lombok.extern.java.Log;

import com.navercorp.scavenger.javaagent.collecting.CodeBaseScanner;
import com.navercorp.scavenger.javaagent.collecting.InvocationTracker;
import com.navercorp.scavenger.javaagent.model.CodeBase;
import com.navercorp.scavenger.javaagent.model.Config;
import com.navercorp.scavenger.javaagent.publishing.Publisher;
import com.navercorp.scavenger.model.CodeBasePublication;
import com.navercorp.scavenger.model.GetConfigResponse;
import com.navercorp.scavenger.model.InvocationDataPublication;

@Log
public class Scheduler implements Runnable {
    private final InvocationRegistry invocationRegistry;
    private final CallStackRegistry callStackRegistry;
    private final Config config;
    private final Publisher publisher;
    private final CodeBaseScanner codeBaseScanner;
    private final ScheduledExecutorService executor;

    private final SchedulerState pollState;
    private final SchedulerState codeBasePublisherState;
    private final SchedulerState invocationDataPublisherState;
    private final SchedulerState callStackDataPublisherState;

    private final int forceIntervalSeconds;
    private final int maxMethodsCount;

    private GetConfigResponse dynamicConfig;
    private CodeBasePublication codeBasePublication;
    private boolean isCodeBasePublished = false;
    private InvocationDataPublication invocationDataPublication;
    private CallStackDataPublication callStackDataPublication;
    private String codeBaseFingerprint;

    public Scheduler(InvocationRegistry invocationRegistry, CallStackRegistry callStackRegistry, Config config) {
        this(invocationRegistry, callStackRegistry, config, new Publisher(config), new CodeBaseScanner(config));
    }

    public Scheduler(InvocationRegistry invocationRegistry, CallStackRegistry callStackRegistry, Config config, Publisher publisher, CodeBaseScanner codeBaseScanner) {
        this.invocationRegistry = invocationRegistry;
        this.callStackRegistry = callStackRegistry;
        this.config = config;
        this.codeBaseScanner = codeBaseScanner;
        this.publisher = publisher;
        this.executor = Executors.newScheduledThreadPool(
            1,
            ScavengerThreadFactory.builder().
                name("scheduler").
                relativePriority(-1)
                .build()
        );

        this.forceIntervalSeconds = this.config.getForceIntervalSeconds();
        this.maxMethodsCount = this.config.getMaxMethodsCount();
        // these intervals will be updated when dynamic config is polled (not likely be used)
        int intervalSeconds = 600;
        int retryIntervalSeconds = 600;

        this.pollState = new SchedulerState("config_poll").initialize(intervalSeconds, retryIntervalSeconds);
        this.codeBasePublisherState = new SchedulerState("codebase").initialize(intervalSeconds, retryIntervalSeconds);
        this.invocationDataPublisherState = new SchedulerState("invocation_data").initialize(intervalSeconds, retryIntervalSeconds);
        this.callStackDataPublisherState = new SchedulerState("call_stack_data").initialize(intervalSeconds, retryIntervalSeconds);

        if (forceIntervalSeconds != 0) {
            this.pollState.updateIntervals(forceIntervalSeconds, forceIntervalSeconds);
            this.codeBasePublisherState.updateIntervals(forceIntervalSeconds, forceIntervalSeconds);
            this.invocationDataPublisherState.updateIntervals(forceIntervalSeconds, forceIntervalSeconds);
            this.callStackDataPublisherState.updateIntervals(forceIntervalSeconds, forceIntervalSeconds);
        }
    }

    public void start() {
        executor.scheduleAtFixedRate(
            this,
            config.getSchedulerInitialDelayMillis(),
            config.getSchedulerIntervalMillis(),
            TimeUnit.MILLISECONDS
        );
    }

    public void shutdown() {
        synchronized (executor) {
            log.info("[scavenger] shutting scheduler down");
            executor.shutdown();
            try {
                //noinspection ResultOfMethodCallIgnored
                executor.awaitTermination(10, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                log.fine("[scavenger] shutdown interrupted");
                Thread.currentThread().interrupt();
            }

            if (dynamicConfig != null) {
                codeBasePublisherState.scheduleNow();
                publishCodeBaseIfNeeded();

                invocationDataPublisherState.scheduleNow();
                publishInvocationDataIfNeeded();

                if (config.isCallStackTraceMode()) {
                    callStackDataPublisherState.scheduleNow();
                    publishCallStackDateIfNeeded();
                }
            }
        }
    }

    @Override
    public void run() {
        synchronized (executor) {
            if (executor.isShutdown()) {
                log.info("[scavenger] scheduler is shutting down");
                return;
            }

            try {
                pollDynamicConfigIfNeeded();
                publishCodeBaseIfNeeded();
                publishInvocationDataIfNeeded();
                publishCallStackDateIfNeeded();
            } catch (Throwable t) {
                log.severe("[scavenger] scheduler failure: " + t);
            }
        }
    }

    public void pollDynamicConfigIfNeeded() {
        if (pollState.isDueTime()) {
            try {
                dynamicConfig = publisher.pollDynamicConfig();
                if (forceIntervalSeconds == 0) { // if forceIntervalSeconds is disabled
                    pollState.updateIntervals(
                        dynamicConfig.getConfigPollIntervalSeconds(),
                        dynamicConfig.getConfigPollRetryIntervalSeconds()
                    );
                    codeBasePublisherState.updateIntervals(
                        dynamicConfig.getCodeBasePublisherCheckIntervalSeconds(),
                        dynamicConfig.getCodeBasePublisherRetryIntervalSeconds()
                    );
                    invocationDataPublisherState.updateIntervals(
                        dynamicConfig.getInvocationDataPublisherIntervalSeconds(),
                        dynamicConfig.getInvocationDataPublisherRetryIntervalSeconds()
                    );
                    if (config.isCallStackTraceMode()) {
                        callStackDataPublisherState.updateIntervals(
                            dynamicConfig.getInvocationDataPublisherIntervalSeconds(),
                            dynamicConfig.getInvocationDataPublisherRetryIntervalSeconds()
                        );
                    }
                }
                pollState.scheduleNext();
            } catch (Exception e) {
                log.log(Level.SEVERE, "[scavenger] poll config failed", e);
                if (e instanceof StatusRuntimeException) {
                    Status.Code code = ((StatusRuntimeException)e).getStatus().getCode();
                    if (code == Status.Code.UNAUTHENTICATED) {
                        log.warning("[scavenger] authentication failed. disabling scavenger...");
                        shutdown();
                    }
                }
                pollState.scheduleRetry();
            }
        }
    }

    public void publishCodeBaseIfNeeded() {
        if (!isCodeBasePublished && codeBasePublisherState.isDueTime() && dynamicConfig != null) {
            try {
                if (this.codeBasePublication == null) {
                    boolean scanSuccessful = scanCodeBase();

                    if (!scanSuccessful) {
                        return;
                    }
                }

                this.publisher.publishCodeBase(codeBasePublication);
                this.isCodeBasePublished = true;
                this.codeBaseFingerprint = codeBasePublication.getCommonData().getCodeBaseFingerprint();
                this.codeBasePublication = null; // remove codebase publication reference to let it garbage collected.
            } catch (Exception e) {
                log.log(Level.SEVERE, "[scavenger] codebase publish failed", e);
                codeBasePublisherState.scheduleRetry();
            }
        }
    }

    public boolean scanCodeBase() {
        CodeBase codeBase;
        try {
            codeBase = codeBaseScanner.scan();
        } catch (Throwable e) {
            log.log(Level.SEVERE, "[scavenger] code scanning is failed. Stop codebase publishing.", e);
            return false;
        }

        if (codeBase.getMethods().isEmpty()) {
            log.severe("[scavenger] no methods are found");
            return false;
        } else if (codeBase.getMethods().size() > maxMethodsCount) {
            log.severe("[scavenger] maximum methods count(" + maxMethodsCount + ") exceed: " + codeBase.getMethods().size());
            return false;
        }
        this.codeBasePublication = codeBase.toPublication(config);
        return true;
    }

    public void publishInvocationDataIfNeeded() {
        if (invocationDataPublisherState.isDueTime() && dynamicConfig != null && isCodeBasePublished) {
            try {
                if (invocationDataPublication == null) {
                    invocationDataPublication = invocationRegistry
                        .getPublication(
                            config,
                            codeBaseFingerprint
                        );
                }
                publisher.publishInvocationData(invocationDataPublication);

                invocationDataPublication = null;
                invocationDataPublisherState.scheduleNext();
            } catch (Exception e) {
                if (e instanceof StatusRuntimeException && ((StatusRuntimeException) e).getStatus().equals(Status.UNAVAILABLE)) {
                    log.log(Level.SEVERE, "[scavenger] invocation data publish failed (server unavailable)");
                } else {
                    log.log(Level.SEVERE, "[scavenger] invocation data publish failed", e);
                }
                invocationDataPublisherState.scheduleRetry();
            }
        }
    }

    public void publishCallStackDateIfNeeded() {
        if (config.isCallStackTraceMode() && callStackDataPublisherState.isDueTime() && dynamicConfig != null && isCodeBasePublished) {
            try {
                if (callStackDataPublication == null) {
                    callStackDataPublication = callStackRegistry
                        .getPublication(
                            config,
                            codeBaseFingerprint
                        );
                }
                publisher.publishCallStackData(callStackDataPublication);
                callStackDataPublication = null;
                callStackDataPublisherState.scheduleNext();
            } catch (Exception e) {
                if (e instanceof StatusRuntimeException && ((StatusRuntimeException) e).getStatus().equals(Status.UNAVAILABLE)) {
                    log.log(Level.SEVERE, "[scavenger] call stack data publish failed (server unavailable)");
                } else {
                    log.log(Level.SEVERE, "[scavenger] call stack data publish failed", e);
                }
                callStackDataPublisherState.scheduleRetry();
            }
        }
    }
}
